### ISSUE TITLE
Finetune Gemma model on LPI Dataset

### ISSUE THREAD

**Issue by TobeTek** [neomatrix369/learning-path-index#69]
Using LoRA, QLoRA, or any other fine-tuning approach:

- [ ] Train an LLM model on the LPI Dataset to recommend courses and/or provide descriptions of courses within the dataset.
- [ ] Create a notebook with the training code, including benchmarks, training progress, and relevant plots.
- [ ] Deploy the fine-tuned model on Kaggle, Hugging Face, or a similar model hosting platform.

**Comment by neomatrix369** [comment]
@Foluwa please update this ticket regularly with your findings

**Comment by Foluwa** [comment]
@neomatrix369 I am currently working on including benchmarks, training progress, and relevant plots so that I can properly evaluate the performance, I have added the notebook that does the finetuning in this [commit](https://github.com/neomatrix369/learning-path-index/compare/main...lpi-gemma-model)

**Comment by neomatrix369** [comment]
@Foluwa Can you pls go thru this notebook and see how we have documented and laid out the notebook and try to follow the same ideas for your notebook(s) as well:
- https://www.kaggle.com/code/neomatrix369/kaggle-global-outreach-analysis
- https://www.kaggle.com/code/neomatrix369/nlp-profiler-simple-dataset

**Comment by Foluwa** [comment]
@neomatrix369 Thanks for sharing, I will go through these notebooks and follow their standard

**Comment by TobeTek** [comment]
A notebook of mine with an interactive UI for data exploration:
https://www.kaggle.com/code/tobetek/eda-rsna-abdominal-injury-competition-2023

**Comment by neomatrix369** [comment]
@Foluwa Check this https://stackoverflow.com/questions/60251383/dockerfile-from-platform-option for your Docker macOS M2 related issue

**Comment by Foluwa** [comment]
# Finetuning Gemma model on LPI Dataset


The initial approach was implemented in this [notebook](https://github.com/neomatrix369/learning-path-index/blob/lpi-gemma-model/app/llm-poc-variant-03/lpi_finetune_gemma.ipynb) but some errors were encountered (which are listed below) and I pivoted to this [notebook](https://github.com/neomatrix369/learning-path-index/blob/lpi-gemma-model/app/llm-poc-variant-03/lpi_keras_finetuning.ipynb) most of the implementation were from ["Fine-tuning Gemma 2 model with role-playing dataset"](https://medium.com/@gabi.preda/fine-tuning-gemma-2-model-with-role-playing-dataset-b8ec399a2e17) by Gabriel Preda

## Issues encountered

### 1.  Resource constraints (ensure you run on a GPU)
<img width="841" alt="Screenshot 2024-10-26 at 5 15 34 PM" src="https://github.com/user-attachments/assets/eae636ad-dda6-44a7-a05c-a3952c263401">
<img width="603" alt="Screenshot 2024-10-26 at 4 17 20 PM" src="https://github.com/user-attachments/assets/b7f4dca6-44b4-42a3-b910-399ef42a444c">
<img width="795" alt="Screenshot 2024-10-26 at 4 17 33 PM" src="https://github.com/user-attachments/assets/885769cf-71a7-406e-90fc-4e019930ad62">


### 2.  Permission Issues
<img width="1027" alt="Screenshot 2024-10-26 at 4 19 27 PM" src="https://github.com/user-attachments/assets/087c89fb-057d-426d-9a42-f17728213982">
<img width="745" alt="Screenshot 2024-10-26 at 4 18 05 PM" src="https://github.com/user-attachments/assets/4d8e46d8-eea3-49b3-817f-e9dd886f38f3">


### 3. The evaluation metric does not support "multiclass-multioutput" format
![Screenshot 2024-10-26 at 5 00 18 PM](https://github.com/user-attachments/assets/5f0f0c62-67ee-4986-83ad-a00aa531aa64)


### 4.  Unable to build the model
![image1](https://github.com/user-attachments/assets/9ee3d7af-e9fa-4af5-b208-d3ac055b96f0)


### 5.  Unable to save the model weights
![Screenshot 2024-10-26 at 5 11 14 PM](https://github.com/user-attachments/assets/3a5e3102-6015-4ec7-92d5-9df088dc668f)


### 6. Model performance issue
![Screenshot_2024-10-07_at_12 55 15_AM](https://github.com/user-attachments/assets/f49d782f-cc42-48ac-b4c3-e108ed5cf563)
